config/merge/merge.go

```diff
@@ -183,6 +183,11 @@ func MergeStruct(parent, child reflect.Value) reflect.Value {
 					if childList == fieldName {
 						// case 1: in child config in same list
 						if childItem.Kind() == reflect.Struct {
+							// If HTTP header Value is nil, it means that we should remove the
+							// parent header from the result.
+							if fieldName == "HTTPHeaders" && childItem.FieldByName("Value").IsNil() {
```

Comment:
I don't quite understand what you mean. Can you elaborate on that?

Reply:
If a parent with `{"key": "X-Auth", "value": "foo"}` is merged with a child with `{"key": "X-Auth"}`, the result of the default merging process (without this `fieldName == "HTTPHeaders"` special case) should be a config with `{"key": "X-Auth", "value": null}`.

So, I'm wondering if we can drop the special case here, and instead have `HTTPHeaders.Parse()` just ignore any HTTPHeader with a nil `Value`.  That would probably require loosening some validation, but it might be worthwhile to avoid the special case.

Reply:
It's not true. If a parent with `{"key": "X-Auth", "value": "foo"}` is merged with a child with `{"key": "X-Auth"}`, the result of the default merging process (without this `fieldName == "HTTPHeaders"` special case) is a config with `{"key": "X-Auth", "value": "foo"}`. So, by default it doesn't modify anything.

Reply:
You're right, of course.

I don't like the special case, but a null value in a key-value object feels pretty well-defined and consistent with the general design of the merging rules, so I'm inclined to leave this as it is now.  @jlebon, could you double-check my reasoning here?

Reply:
Hmm, yeah I think we have to leave this as is. Given that we use pointers in a lot of places in the schema to handle optional fields, there would be no way to discern between an explicit null and an omitted field.

Reply:
Chatted with @bgilbert a bit more about this. It's not immediately obvious that `{"key": "My-Header"}` translates to "delete the value", but having an explicit `delete: true` key for this would be a departure as well from the way other fields are merged. (There are no other fields in the spec which are "self-aware" about possibly being merged into lower configs, apart from the `merge` bit itself of course.)

It's also not clear if we did that whether `{"key": "My-Header", "delete": true}` should differ in behaviour from `{"key": "My-Header", "value": "value", "delete": true}`. So in the interest of keeping the spec consistent even if at the expense of consistency in the code, I agree the current behaviour in this code makes sense.